### PR TITLE
Visit artesian spring once a week

### LINKED ISSUE #3470: Adventure map, Artesian Spring bonus should only be given to the first hero, all players taken into account, who visits the Artesian Spring and only once a week
The Artesian Spring bonus should only be given to the first hero, all players taken into account, who visits the Artesian Spring and only once a week.

In the example, Sir Gallant visits the Artesian Spring before Sandro, but both heroes receive the Artesian Spring bonus; only Sir Gallant should receive the bonus because he is the first in the current week to visit the Artesian Spring.

In fheroes2:

Artesian bonus once a week only 001

Artesian bonus once a week only 002

https://user-images.githubusercontent.com/43583994/119190059-220eb680-ba7d-11eb-9ec8-6ba4462bc6b8.mp4

End the turn.

Here are the save files (before and after):
[Artesian Spring bonus 01.zip](https://github.com/ihhub/fheroes2/files/6524761/Artesian.Spring.bonus.01.zip)

### PR DESCRIPTION
**Pull request by tau3**
Fixes #3470 
Only one hero will be able to visit an Artesian Spring in a week.

Fix notes:
Probably there is a more convenient way to iterate through all colors, but I couldn't find it. Maybe `IsAnyKingdomVisited` should be a member of `World` and `Kingdoms` should be an iterable class? It would simplify code.

### CODE REVIEW

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @tau3 , there are multiple comments left in this pull request. Could you please take a look?

File: src/fheroes2/world/world.cpp [hunk 1]
```diff
@@ -971,6 +971,18 @@ bool World::KingdomIsWins( const Kingdom & kingdom, int wins ) const
     return false;
 }
 
+bool IsAnyKingdomVisited( const u32 obj, const s32 dstIndex )
```
[the remark this answers is not in the document]
Comment (tau3) [reply src/fheroes2/world/world.cpp:974]: So the code style has had some changes? I remember changing to capitals earlier

File: src/fheroes2/world/world.cpp [hunk 2]
```diff
@@ -971,6 +971,18 @@ bool World::KingdomIsWins( const Kingdom & kingdom, int wins ) const
...
+{
+    const int maxColor = 1 << KINGDOMMAX;
```
[the remark this answers is not in the document]
Comment (tau3) [reply src/fheroes2/world/world.cpp:976]: yeah, this one is better, thank you

File: .gitignore [hunk 3]
```diff
@@ -75,3 +75,14 @@ VisualStudio/packages/installed
 
 # Mac magic folders
 .DS_Store
+
```
Comment (ihhub) [inline .gitignore:78]: Could we please revert this change as it doesn't belong to the issue? It's good to separate changes

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @tau3 , just one more comment. Would you mind to have a look?

File: src/fheroes2/world/world.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/world/world.cpp]: Most of the new code is written for functions starting with non-capital letter :)

**Comment by ihhub** [comment]
@tau3 , thank you very much for this fix!

File: .gitignore (hunk 3 above)
Comment (tau3) [reply .gitignore]: ok, i'll push those changes to `.gitignore` in a separate one

**Pull request merged**